### PR TITLE
Gjoshi/implement token revoke experience

### PR DESCRIPTION
**Pull request by goravjoshi**
This PR covers the following work,
1. The authorize servlet which redirects the consumer back to the source once the user grants the request. 
2. UI as described below as a screenshots for an idea (Might be slightly older).

### Jenkins Management Option
image

### OAuth Consumer List
image

### Consumer Details
image

### Add New Consumer
image

### Access Token List
image

### Authorize UI for the end-user
image

### CODE REVIEW

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/applink/oauth/serviceprovider/consumer/PersistentServiceProviderConsumerStore.java [hunk 1]
```diff
@@ -92,6 +92,17 @@ public void delete(String key) {
         }
     }
 
+    @Override
```
Comment (alibk0rd) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/applink/oauth/serviceprovider/consumer/PersistentServiceProviderConsumerStore.java]: Sorry, forgot to add this. Was a crazy couple of weeks. :man_facepalming:

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookTriggerImpl.java [hunk 2]
```diff
@@ -136,7 +136,7 @@ boolean isWorkflowJob(@Nullable SCMTriggerItem triggerItem) {
      *
      * For pipeline job, this is invoked every time a build is run.
      *
-     * We can't always continue registration check if newInstance is false since during Jenkin startup, this is invoked
+     * We can't always continue registration check if createInstance is false since during Jenkin startup, this is invoked
```
Comment (alibk0rd) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookTriggerImpl.java]: was this intentional? Looks like IDEA trying to be too smart. ;-)

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/jenkins/oauth/servlet/AuthorizeAction.java [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+import java.util.Map;
+import java.util.logging.Logger;
+
+import static com.atlassian.bitbucket.jenkins.internal.applink.oauth.serviceprovider.token.ServiceProviderToken.Authorization;
+import static javax.servlet.http.HttpServletResponse.SC_UNAUTHORIZED;
+import static jenkins.model.Jenkins.ANONYMOUS;
+import static net.oauth.OAuth.*;
+import static net.oauth.OAuth.Problems.*;
+
+public class AuthorizeAction extends AbstractDescribableImpl<AuthorizeAction> implements Action {
```
Comment (mhenschke-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/jenkins/oauth/servlet/AuthorizeAction.java]: This class was named "AuthorizeAction" to try to fit in line with existing Jenkins naming conventions, but these are not being observed elsewhere for actions- we seem to be using the terms "Configuration" and "Endpoint". If we want to keep using those conventions, this one should probably be renamed "ConsumerTokenConfiguration" or something similar.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/jenkins/oauth/token/DisplayAccessToken.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    public DisplayAccessToken(ServiceProviderToken token, Clock clock) {
+        this.token = token;
+        this.clock = clock;
+    }
+
+    @SuppressWarnings("unused") // Stapler
+    public Date getCreationDate() {
+        return new Date(token.getCreationTime());
+    }
+
```
Comment (mhenschke-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/jenkins/oauth/token/DisplayAccessToken.java]: /nit 
```suggestion
@SuppressWarnings("unused") // Stapler
```

**Comment by mhenschke-atl** [comment]
The only thing I really noticed is there is no form validation in the `OAuthCreateAction`, but if this is not expected to be used as part of the applink experience in the end result, it's not a priority to fix.
Besides that, just or or two tiny nits. LGTM.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/applink/oauth/serviceprovider/consumer/PersistentServiceProviderConsumerStore.java [hunk 5]
```diff
@@ -92,6 +92,17 @@ public void delete(String key) {
...
+    public void update(Consumer consumer) {
+        requireNonNull(consumer, "consumer");
+        load();
+        if (!entityMap.containsKey(consumer.getKey())) {
+            log.warning(() -> String.format("Consumer with key '%s' does not exists.", consumer.getKey()));
+            throw new StoreException("No Consumer with the given key" + consumer.getKey());
```
[the remark this answers is not in the document]
Comment (goravjoshi) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/applink/oauth/serviceprovider/consumer/PersistentServiceProviderConsumerStore.java]: Can you also share the reason why? In my opinion, if I have to look for a wrong key name, I will see it through UI and then logs. Having it in the exception message propagates it to the correct place.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/jenkins/oauth/servlet/AuthorizeAction.java (hunk 3 above)
Comment (goravjoshi) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/jenkins/oauth/servlet/AuthorizeAction.java]: `Action` was used in the general sense and not in Jenkins sense. This is when user authorizes. I can rename it to `AuthorizeConfirmationPage` if that is more clear. What do you think?

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookTriggerImpl.java (hunk 2 above)
Comment (goravjoshi) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookTriggerImpl.java]: Yeah, IntelliJ super sensitive refactor. I'll fix it.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/jenkins/oauth/servlet/AuthorizeAction.java [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+
+    //Following fields are used in Jelly file
+    public static final String AUTHORIZE_KEY = "authorize";
+    public static final String CANCEL_KEY = "cancel";
+    public static final String OAUTH_TOKEN_PARAM = "oauth_token";
+
```
[the remark this answers is not in the document]
Comment (goravjoshi) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/jenkins/oauth/servlet/AuthorizeAction.java]: A blank line is to separate `public static` from `private static`.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/jenkins/oauth/servlet/AuthorizeAction.java (hunk 3 above)
Comment (mhenschke-atl) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/jenkins/oauth/servlet/AuthorizeAction.java]: The `Action` label appears fairly widely (though also at times inconsistently) in the Jenkins code base, so if we're not using it, I think we should avoid it altogether. I'm happy with `AuthorizeConfirmationPage` (or maybe `AuthorizeConfirmationForm`, which we could apply to any page with a form component?)

**Pull request merged**